llvm-general/src/LLVM/General/Internal/Module.hs

```diff
@@ -80,20 +96,22 @@ genCodingInstance [t| Bool |] ''FFI.LinkerMode [
   (FFI.linkerModeDestroySource, False)
  ]
 
--- | link LLVM modules - move or copy parts of a source module into a destination module.
--- Note that this operation is not commutative - not only concretely (e.g. the destination module
--- is modified, becoming the result) but abstractly (e.g. unused private globals in the source
--- module do not appear in the result, but similar globals in the destination remain).
+-- | link LLVM modules - move or copy parts of a source module into a
+-- destination module.  Note that this operation is not commutative -
+-- not only concretely (e.g. the destination module is modified,
+-- becoming the result) but abstractly (e.g. unused private globals in
+-- the source module do not appear in the result, but similar globals
+-- in the destination remain). The source module is destroyed.
 linkModules ::
-  Bool -- ^ True to leave the right module unmodified, False to cannibalize it (for efficiency's sake).
-  -> Module -- ^ The module into which to link
-  -> Module -- ^ The module to link into the other (and cannibalize or not)
+     Module -- ^ The module into which to link
+  -> ModuleRef -- ^ The module to link into the other (and cannibalize or not)
   -> ExceptT String IO ()
-linkModules preserveRight (Module m) (Module m') = flip runAnyContT return $ do
-  preserveRight <- encodeM preserveRight
-  msgPtr <- alloca
-  result <- decodeM =<< (liftIO $ FFI.linkModules m m' preserveRight msgPtr)
-  when result $ throwError =<< decodeM msgPtr
+linkModules (Module dest) srcRef = flip runAnyContT return $ do
+  (Module src) <- liftIO $ getModule srcRef
+  result <- decodeM =<< liftIO (FFI.linkModules dest src)
+  -- linkModules takes care of deleting the sourcemodule
+  liftIO $ deleteModule srcRef
+  when result (throwError "Couldn’t link modules")
```

Comment:
Oh yuck. The real messages have been shuffled into state hanging off the context. It's pretty ugly to have a signature promising error messages and then deliver only a made up "didn't work" string.

We could expose the diagnostic handler in the context and give this function a simpler signature. Significantly better, I think, would be to implement the diagnostic handler ourselves and return the actual errors here.